Generate snapshot in the CI

TeddyAndrieux opened the pull request:
**Component**:

'eve',
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

#2906 

**Summary**:

- Make multi-node stage handling various number of nodes (up to 2 for the moment as we do not handle more in pytest)
- Add server IDs in terraform outputs so that we can use those IDs to directly act on server using `openstack` client
- Add a script to generate snapshot from terraform output using `openstack` client
- Add a stage to generate 3 nodes and single node snapshots

---

Fixes: #2906

TeddyAndrieux committed:
eve: Add stage to generate snapshot

Add a stage that use `multi-node` and `single-node` stages to
generate snapshot that will be used later for upgrade test

Fixes: #2906

TeddyAndrieux committed:
eve: Add a script to generate snapshot from terraform output

Add a simple bash script that use terraform environment variable and
terraform output to generate snapshot with `openstack` client

Refs: #2906

TeddyAndrieux committed:
eve: Add all VM Ids in terraform output

Add all server openstack ids in terraform output so that it can be used
to do some action directly on the server using `openstack` client

TeddyAndrieux committed:
eve: Make multi-node and single-node stage more generic

Multi-node stage could be use to do an expansions of various number of
node, so add the support from eve to generate "numbers" of node using
property as input.
Single-node and multi-node stage could be use to do installation with
different OS, so add the support from eve to install different OS using
property as input.

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

TeddyAndrieux committed:
eve: Make multi-node and single-node able to install Solution

Add some step in multi-node and single-node stage so that (if enabled
with a property) we install a solution on the resulting MetalK8s
environment

TeddyAndrieux committed:
eve: Include version in ISO file name when upgrading using snapshot

When we upgrade from snapshot the `metalk8s.iso` file may already exists
in the snapshot, so to avoid removing/fail during upgrade stage because
of this, rename the file on the bootstrap node to
`metalk8s-<version>.iso`

TeddyAndrieux commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

TeddyAndrieux committed:
eve: Allow redirection when downloading ISOs

Allow url redirection when downloading ISOs in eve stages

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/generate-snapshot-in-ci` into
`development/2.7`
* [w/2.8/improvement/generate-snapshot-in-ci](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/improvement/generate-snapshot-in-ci) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #3b4102457471c390101c8bc18bde02d69b8f3488 on the integration branch
`w/2.8/improvement/generate-snapshot-in-ci` is merging a branch which is neither the current
branch `improvement/generate-snapshot-in-ci` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `improvement/generate-snapshot-in-ci` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #37464a782fd70f6c29f7e7ae6631dc808ad99761 on the integration branch
`w/2.8/improvement/generate-snapshot-in-ci` is merging a branch which is neither the current
branch `improvement/generate-snapshot-in-ci` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `improvement/generate-snapshot-in-ci` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/generate-snapshot-in-ci` into
`development/2.7`
* [w/2.8/improvement/generate-snapshot-in-ci](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/improvement/generate-snapshot-in-ci) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

thomasdanan reviewed (approved):
Snapshot present when launching the eve stage ...

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

gdemonet commented on eve/main.yml:
```diff
@@ -597,11 +601,20 @@ models:
+        REPORT_OWNER: ""
+        REPORT_GROUP: ""
+      command: |-
+        HOSTS_LIST=${HOSTS_LIST:-bootstrap$(seq -s '' --format ' node-%%g' 1 %(prop:nodes_count:-0)s)}
```
(really minor) IIUC `-s` describes a separator, so why not use `seq -s ' ' --format 'node-%g' 1 $nodes_count` for clarity?

TeddyAndrieux replied:
Just because I do not want to have "bootstrap " if we have no nodes :smile:

gdemonet replied:
It doesn't seem to be the behaviour of `seq`:

```bash
$ val=$(seq -s ' ' --format 'node-%g' 1 0)
$ echo "<start>$val<end>"
<start><end>
```

TeddyAndrieux replied:
Yes exactly but  "bootstrap" is not part of this `seq` 
So if you do 
```
# echo -e "<start>bootstrap $(seq -s ' ' --format 'node-%g' 1 0)<end>"
<start>bootstrap <end>

gdemonet replied:
Ahhh, right, :man_facepalming:

gdemonet commented on eve/main.yml:
```diff
@@ -2073,45 +2086,44 @@ stages:
     steps:
       - Git: *git_pull
+      - SetProperty:
+          # NOTE: Set property will be skipped if the value is already set,
```
Always? Skipped if set by user or also by some "parent build stage"?

TeddyAndrieux replied:
Btw it's not the normal behavior of "SetProperty" but since I do a `value: "%(prop:nodes_count:-2)s"` yes it's always the case

Or yes you right it's not really "skipped" just we take the current value. I will update the comment message in another PR

gdemonet replied:
> since I do a value: "%(prop:nodes_count:-2)s" yes it's always the case

Ah right I missed this :+1: no rush for updating the comment :wink:

gdemonet commented on eve/main.yml:
```diff
@@ -2042,7 +2055,7 @@ stages:
           env:
             <<: *_env_bastion_tests
-            PYTEST_FILTERS: "install and ci and multinodes and not node2"
+            PYTEST_FILTERS: "install and ci and multinodes and not node$((%(prop:nodes_count)s + 1))"
```
I guess we'll need to adjust these "tests", don't think they can handle more than 2 nodes.

TeddyAndrieux replied:
Yes but we do not need more than 2 nodes for the moment and since it's only internal I think it's ok but you right if one day we need more we will need to update the expansion "test" (or change the way we do expansion in the CI :wink: )

gdemonet commented on eve/main.yml:
```diff
@@ -589,6 +591,83 @@ models:
+      doStepIf: "%(prop:install_solution:-false)s"
+      name: Activate Solution
+      env:
+        SOLUTION_NAME: "%(prop:solution_name:-example-solution)s"
+        SOLUTION_VERSION: "%(prop:solution_version:-1.0.0)s"
```
Can't we read these from the downloaded archive? This could serve as a minimal check, and would reduce the number of build props to set (using a Zenko archive as an example). The current approach needs at least 5 properties to be set:

  - `install_solution=true`
  - `solution_base_url=https://zenko-artifacts/2.3.4/` (or `solution_repo=https://zenko-artifacts/`)
  - `solution_archive=zenko-2.3.4.iso`
  - `solution_name=zenko`
  - `solution_version=2.3.4`

Reading the name and version from the ISO could reduce the required props to `install_solution`, `solution_base_url` and `solution_archive` (the two latter could be reduced to a single `solution_archive_url` from which we can extract the basename).

TeddyAndrieux replied:
Hmm you right, I will do it, in another PR

gdemonet reviewed (approved):
Some minor comments (which we can address later), overall LGTM :+1:

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
